internal/jimm/juju/jimm.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2025 Canonical.
+package juju
+
+import (
+	"context"
+	"database/sql"
+	"strings"
+
+	"github.com/google/uuid"
+	jujuparams "github.com/juju/juju/rpc/params"
+	"github.com/juju/names/v5"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/jimm/credentials"
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+var (
+	initiateMigration = func(ctx context.Context, j *JIMM, user *openfga.User, spec jujuparams.MigrationSpec) (jujuparams.InitiateMigrationResult, error) {
+		return j.InitiateMigration(ctx, user, spec)
+	}
+)
+
+// forEachController runs a given function on multiple controllers
+// simultaneously. A connection is established to every controller in the
+// given list concurrently and then the given function is called with the
+// controller and API connection to use to perform the controller
+// operation. ForEachConnection waits until all operations have finished
+// before returning, any error returned will be the first error
+// encountered when connecting to the controller or returned from the given
+// function.
+func (j *JIMM) forEachController(ctx context.Context, controllers []dbmodel.Controller, f func(*dbmodel.Controller, API) error) error {
+	eg := new(errgroup.Group)
+	for i := range controllers {
+		i := i
+		eg.Go(func() error {
+			api, err := j.dial(ctx, &controllers[i], names.ModelTag{})
+			if err != nil {
+				return err
+			}
+			defer api.Close()
+			return f(&controllers[i], api)
+		})
+	}
+	return eg.Wait()
+}
+
+// ControllerInfo returns info about a controller connected to JIMM.
+func (j *JIMM) ControllerInfo(ctx context.Context, name string) (*dbmodel.Controller, error) {
+	const op = errors.Op("jimm.ListControllers")
+	ctl := dbmodel.Controller{
+		Name: name,
+	}
+	if err := j.Database.GetController(ctx, &ctl); err != nil {
+		return nil, errors.E(op, err)
+	}
+	return &ctl, nil
+}
+
+// ListControllers returns a list of controllers the user has access to.
+func (j *JIMM) ListControllers(ctx context.Context, user *openfga.User) ([]dbmodel.Controller, error) {
```

Comment:
i think that logic that governs JIMM's concerns should live in a manager of its own..  e.g. most things that are used in the JIMM facade.. what do you think?

Reply:
I don't think it really makes sense in the way we've split things up. Our domains are based on resources, e.g. `groups`, `roles`, etc. The juju package contains what could be `models`, `controllers`, etc, domains. A jimm package would have cross cutting concerns with out packages.